lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/StateMachine/SystemStateMachine.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+// 
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.Driver.StateMachine
+{
+    /// <summary>
+    /// This is a state transition machine which wraps current state, defines valid state transitions and API to move to next state. 
+    /// All the system states are define in <see cref="SystemState"></see>
+    /// For the system state transition diagram <see href="https://issues.apache.org/jira/browse/REEF-1335"></see>
+    /// </summary>
+    internal sealed class SystemStateMachine
+    {
+        internal readonly static IDictionary<StateTransition<SystemState, SystemStateEvent>, SystemState> Transitions = 
+            new ReadOnlyDictionary<StateTransition<SystemState, SystemStateEvent>, SystemState>(
+                new Dictionary<StateTransition<SystemState, SystemStateEvent>, SystemState>()
+        {
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.WaitingForEvaluator, SystemStateEvent.AllContextsAreReady), SystemState.SubmittingTasks },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.WaitingForEvaluator, SystemStateEvent.NotRecoverable), SystemState.Fail },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.AllTasksAreRunning), SystemState.TasksRunning },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.FailedNode), SystemState.ShuttingDown },
+            { new StateTransition<SystemState, SystemStateEvent>(SystemState.SubmittingTasks, SystemStateEvent.FailedNode), SystemState.ShuttingDown },
```

Comment:
Pairs of transitions for FailedNode on lines 37-38 and 39-40 are duplicated, we need only one copy of each.